File: app/src/main/java/com/polok/eubmanagement/presentation/notice/noticelist/NoticeListFragment.java
package com.polok.eubmanagement.presentation.notice.noticelist;

import android.os.Bundle;
import android.view.View;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import com.polok.eubmanagement.R;
import com.polok.eubmanagement.base.BaseFragment;
import com.polok.eubmanagement.base.BaseViewModel;
import com.polok.eubmanagement.databinding.FragmentNoticeListBinding;
import com.polok.eubmanagement.widget.PrimaryLoader;

public class NoticeListFragment extends BaseFragment<FragmentNoticeListBinding> {
    @Override
    protected FragmentNoticeListBinding initViewBinding() {
        return FragmentNoticeListBinding.inflate(getLayoutInflater());
    }
    private NoticeListViewModel viewModel;
    @Override
    protected BaseViewModel initViewModel() {
        return viewModel = new ViewModelProvider(this).get(NoticeListViewModel.class);
    }
    private NoticeListAdapter noticeListAdapter;
    @Override
    protected void initOnCreateView(Bundle savedInstanceState) {

        viewModel.fetchNoticeListFromFirebase();

        viewModel.getNoticeLiveData().observe(getViewLifecycleOwner(), noticeList-> {
            if (noticeList != null && !noticeList.isEmpty()) {
                noticeListAdapter = new NoticeListAdapter(noticeList ,new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        viewModel.navigateToViewNoticeFragment(noticeList.get(noticeListAdapter.getAdapterPosition()));
                    }
                });
                binding.noticeRecycler.setAdapter(noticeListAdapter);
            }
        });
        binding.addNoticeButton.setOnClickListener(view -> {
            viewModel.fireNavigateEvent(R.id.action_viewNoticeFragment_to_addNoticeFragment,null);
        });
    }
    @Override
    protected PrimaryLoader initPrimaryLoader() {return binding.primaryLoader;}
    @Override
    protected void onNavigateEvent(int id, Bundle bundle) {
        super.onNavigateEvent(id, bundle);
        Navigation.findNavController(binding.getRoot()).navigate(id, bundle);
    }
}

File: app/src/main/java/com/polok/eubmanagement/presentation/notice/noticelist/NoticeListFragment.kt
package com.polok.eubmanagement.presentation.notice.noticelist

import android.os.Bundle
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.polok.eubmanagement.R
import com.polok.eubmanagement.base.BaseFragment
import com.polok.eubmanagement.base.BaseViewModel
import com.polok.eubmanagement.databinding.FragmentNoticeListBinding
import com.polok.eubmanagement.util.SharedPref
import com.polok.eubmanagement.util.makeVisible
import com.polok.eubmanagement.widget.PrimaryLoader

class NoticeListFragment : BaseFragment<FragmentNoticeListBinding>(
    viewBindingFactory = FragmentNoticeListBinding::inflate
) {
    private val viewModel: NoticeListViewModel by viewModels {
        NoticeListViewModel.Factory()
    }
    private val adapter: NoticeListAdapter by lazy {
        NoticeListAdapter { noticeData ->
            viewModel.navigateToViewNoticeFragment(noticeData)
        }
    }

    override fun initViewModel(): BaseViewModel = viewModel

    override fun initOnCreateView(savedInstanceState: Bundle?) {

        viewModel.fetchNoticeListFromFirebase()

        adapter.isAdmin = SharedPref.getUserProfile().isAdmin
        if (adapter.isAdmin == true) binding.addNoticeButton.makeVisible()

        viewModel.noticeLiveData.observe(viewLifecycleOwner) {
            if (it?.isNotEmpty() == true) {
                binding.noticeRecycler.adapter = adapter.apply {
                    submitList(it)
                }
            }
        }

        binding.addNoticeButton.setOnClickListener {
            viewModel.fireNavigateEvent(
                R.id.action_viewNoticeFragment_to_addNoticeFragment, null
            )
        }
    }

    override fun initPrimaryLoader(): PrimaryLoader = binding.primaryLoader

    override fun onNavigateEvent(id: Int, bundle: Bundle?) {
        super.onNavigateEvent(id, bundle)
        findNavController().navigate(id, bundle)
    }
}